File: backend-java/src/main/java/com/bffgin/backend/auth/ExternalAuth.java
package com.bffgin.backend.auth;

import com.bffgin.backend.domain.TaskError;

/**
 * 外部公開API(CONTRACT.mdセクション11)向けの認証。内部REST/gRPCと同じDispatcherで
 * 署名検証まで行うが、ここではさらに2点を追加で要求する(backend-c/backend-cpp/backend-rustの
 * RequireExternalClientAuthと同じ設計):
 *   1. ローカル(HMAC/RSA)発行のissuerは拒否する(Client Credentials Grant、
 *      つまりKeycloak発行のトークンのみを受け付ける)
 *   2. `azp`(authorized party)クレームが期待するクライアントidと一致すること
 * user_id自体はこのクレームから解決せず、呼び出し側がクエリパラメータで指定した値を
 * そのまま信頼する(サーバー間の信頼関係を前提にした設計、CONTRACT.md参照)
 */
public final class ExternalAuth {

    private ExternalAuth() {
    }

    public static void requireExternalClient(Dispatcher dispatcher, String authorizationHeader,
            String externalApiClientId) throws TaskError {
        if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
            throw TaskError.unauthorized();
        }
        String token = authorizationHeader.substring("Bearer ".length());

        Claims claims;
        try {
            claims = dispatcher.verify(token);
        } catch (VerifyException e) {
            throw TaskError.unauthorized();
        }

        if (Dispatcher.isLocalIssuer(claims.iss())) {
            throw TaskError.unauthorized();
        }
        if (!externalApiClientId.equals(claims.azp())) {
            throw TaskError.unauthorized();
        }
    }
}

File: backend-kotlin/src/main/kotlin/com/bffgin/backend/auth/ExternalAuth.kt
package com.bffgin.backend.auth

import com.bffgin.backend.rest.ExternalApiError

/**
 * 外部公開API専用の認証(backend(Go)のRequireExternalClientAuth・backend-c/backend-cppの
 * auth_require_external_client/RequireExternalClientAuth相当)。内部REST/gRPCのUserResolverとは
 * 別の判定基準を使う:
 *   1. まず通常のDispatcherで署名検証(issで振り分け、実際の検証は委譲先のVerifierが行う)
 *   2. ローカル(HMAC/RSA)発行のissは拒否する(Client Credentials Grant、つまりKeycloak発行分のみ許可)
 *   3. azp(authorized party)クレームがexternalApiClientIdと一致することを要求する
 * user_idはクエリパラメータをそのまま使い、JWTのsubとは突き合わせない
 * (この資格情報を持つ者は任意ユーザーのタスクを読み取れる、CONTRACT.mdセクション11に明記された既知の設計)
 */
object ExternalAuth {
    suspend fun requireExternalClient(
        authorizationHeader: String?,
        dispatcher: Dispatcher,
        externalApiClientId: String,
    ): Claims {
        if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
            throw ExternalApiError.unauthenticated()
        }
        val token = authorizationHeader.substring("Bearer ".length)

        val claims = try {
            dispatcher.verify(token)
        } catch (e: VerifyException) {
            throw ExternalApiError.unauthenticated()
        }

        if (Dispatcher.isLocalIssuer(claims.iss)) {
            throw ExternalApiError.unauthenticated()
        }
        if (claims.azp != externalApiClientId) {
            throw ExternalApiError.unauthenticated()
        }
        return claims
    }
}
